[HIBIKE] create hibike simulator for runtime integration

karthik-shanmugam opened the pull request:
- rewrite `hibike_simulator.py` and add a function `hibike_process`, which takes in various Queues and Pipes and simulates the behavior described in `ATALANDA_INTEGRATION.md`

- rewrite `README.md`, `hibikeDevices.json`, and `hibike_message.py` with protocol changes

- add an `__init__.py` so hibike can be imported from runtime

karthik-shanmugam committed:
create hibike simulator for runtime integration

yizow commented on hibike/ATALANTA_INTEGRATION.md:
```diff
@@ -0,0 +1,83 @@
+# Hibike Runtime Integration
```
Rename the file to just RUNTIME_INTEGRATION. 

Atalanta is a deprecated name

yizow commented on hibike/ATALANTA_INTEGRATION.md:
```diff
@@ -0,0 +1,83 @@
+
+MetalDetector
+
+- callibrate: bool write only
```
typo

nikitakit commented:
I see that I have been assigned. Reviewing this now, but given the size of the change it may take a while.

nikitakit commented on hibike/README.md:
```diff
@@ -46,75 +104,118 @@ Each Smart Device will be assigned an 88-bit UID with the following data.
+
+- Year
+  - 8-bit ID corresponding to the competition year that the Smart Device was manufactured for. 
+  - The 2015-2016 season will correspond to 0x00
```
The spec should be updated with a new year, `0x01`.

Additionally, is possible to detect last year's hibike devices and warn about protocol incompatibility to students? I would not be at all surprised if students try to plug in last year's sensors to their robots. Breaking compatibility is fine as long this does not become a yearly thing, but having old sensors crash the robot would be unfortunate.

nikitakit commented on hibike/README.md:
```diff
@@ -46,75 +104,118 @@ Each Smart Device will be assigned an 88-bit UID with the following data.
-    +------------------------------------+
+|   ID    |       Message Type       |
+|---------|--------------------------|
+|  0x10   |          Ping            |
```
Have you thought carefully about whether future kits will be able to extend hibike in a backward-compatible way? Breaking changes this year is fine, but ideally we could use a few years of experience building hibike to then lock in a stable system.

karthik-shanmugam replied:
Future kits can add new message types; this year's devices will simply ignore them

nikitakit commented on hibike/README.md:
```diff
@@ -46,75 +104,118 @@ Each Smart Device will be assigned an 88-bit UID with the following data.
+|         |                | 4            | s2         | bool       | yes   | yes    |
+|         |                | 5            | s3         | bool       | yes   | yes    |
+|         |                | 6            | s3         | bool       | yes   | yes    |
+|  0x06   | Grizzly        |              |            |            |       |        |
```
What about Yogi Bear? Does it get an ID?

nikitakit commented on hibike/README.md:
```diff
@@ -124,187 +225,152 @@ Note: As of now, Grizzlies are not supported by Hibike (pyGrizzly should
-    |   0xFC  |   Malformed   |
-    |         |    Message    |
-    +-------------------------+
-    |   0xFD  |  Invalid UID  |
```
I'm seeing some of these error types disappear. Is that intentional?

karthik-shanmugam replied:
Yes, they were never actually implemented last year.

nikitakit commented on hibike/README.md:
````diff
@@ -46,75 +104,118 @@ Each Smart Device will be assigned an 88-bit UID with the following data.
+
+    ```
+    +-------------------+-----------------------------+     +-----------------------------+
+    |      Params       |         Value 0             | ... |         Value 15            |
````
Let's say you only request 2 params via the bitmask. Does the value section only contain data for two values, or will there be 16 value sections (with most of them set to zero)? I think this aspect of the spec can be clarified.

karthik-shanmugam replied:
It will only contain data for the two requested via bitmask

nikitakit commented on hibike/__init__.py:
```diff
@@ -0,0 +1,3 @@
+__all__ = ['hibike_simulator', 'hibike_message']
+from . import hibike_simulator
+from . import hibike_message
```
Note the little red symbol at the last line of this diff. It means you don't have a newline at the end of file. This is typically discouraged style, because it pops up in the diff like this.

karthik-shanmugam replied:
Resolved

nikitakit commented on hibike/hibikeDevices.json:
```diff
@@ -1,288 +1,83 @@
-}
\ No newline at end of file
+[
+    {"id": 0,     "name": "LimitSwitch",   "params": [    
```
BTW, the formatting here is super awkward in diff view. Is there any change you could adjust the whitespace in this file?

yizow replied:
This is a json file right, whitespace doesn't have semantic meaning? Just do indents normally. Currenyl, the "number": 0 stuff is waaaaay indented in.

nikitakit replied:
(I would still prefer this to be changed)

yizow replied:
Yeah, same. My comment was a suggestion on how to change it.

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -1,55 +1,74 @@
-  messageTypes["DeviceStatus"] :         5,
-  messageTypes["DeviceResponse"] :       5,
-  messageTypes["Error"] :                1
+  "LimitSwitch"   :       0x00,
```
Is there a reason these device IDs are inlined into this file, when the JSON could be sourced instead?

karthik-shanmugam replied:
There was no reason; they've been removed now

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -106,57 +125,203 @@ def checksum(data):
-  """ Makes and returns DeviceStatus message."""
-  temp_payload = struct.pack('<B', param)
+def decode_params(device_id, params):
+#     Decodes an inputted set of parameters that is in binary form   
```
Docstring instead of comment for the docs

nikitakit commented on hibike/ATALANTA_INTEGRATION.md:
```diff
@@ -0,0 +1,83 @@
+
+## StateManager -> Hibike
+
+`["ready"]`
```
I would recommend renaming this to "init_hibike" or similar.

nikitakit commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,144 @@
+
+
+def hibike_process(badThingsQueue, stateQueue, pipeFromChild):
+    while pipeFromChild.recv()[0] != "ready":
```
(repeating my recommendation to rename "ready")

yizow replied:
The ready command is actually deprecated and this line should be removed.

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -106,57 +125,203 @@ def checksum(data):
     chk ^= data[i]
   return chk
 
+
```
Note that I didn't actually review the checksum for correctness.

nikitakit reviewed (commented):
I've left my comments inline.

I think it would be useful to have someone who worked a lot with last year's hibike protocol spec review this year's update. I never really delved into the nitty-gritty details, so I may have missed something.

yizow reviewed (changes requested):
Haven't fully reviewed. 
Just a couple things I noticed on a first cursory pass.
Will do another review.

karthik-shanmugam committed:
Update and rename ATALANTA_INTEGRATION.md to RUNTIME_INTEGRATION

remove deprecated ready command and fix typo

karthik-shanmugam committed:
Update README.md

- Clarify structure of params and values
- Update year
- Rename Grizzly to YogiBear

karthik-shanmugam committed:
add newline to end of  __init__.py

karthik-shanmugam committed:
clean up hibike_message.py

- remove unnecessary deviceTypes dictionary
- replace comment with docstring

karthik-shanmugam committed:
remove ready command from hibike_simulator.py

yizow commented on hibike/hibikeDevices.json:
```diff
@@ -1,288 +1,83 @@
\ No newline at end of file
+[
+    {"id": 0,     "name": "LimitSwitch",   "params": [    
+                                                          {"number": 0  , "name": "switch0"    , "type": "bool"    , "read": true , "write": false },
```
If we're interested in efficiency, read/ write can be represented as a single integer 0, 1, 2, 3, representing none, either, or both. None doesn't really make sense though.

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
-        return 0
+import threading
+import multiprocessing
+from . import hibike_message as hm
```
Im still confused why we need the .

karthik-shanmugam replied:
hibike_message isn't in the python path when you try importing this from the runtime directory

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+
+
+def hibike_process(badThingsQueue, stateQueue, pipeFromChild):
+    ports = ['0', '1']
```
why not just delete ports, and use serials = range(2) for everythinG?

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+    serials = [int(port) for port in ports]
+
+    # each device has it's own write thread, with it's own instruction queue
+    instruction_queues = [queue.Queue() for _ in ports]
```
How many things are going to be writing to this queue? I thought hibike had a centralized model, meaning there is only one thread putting things on here. A pipe would be faster.

karthik-shanmugam replied:
This is not an IPC Queue; it's just a thread safe Queue data structure.

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+fake_uids = [0 << 72, 7 << 72]
+
+
+uid_to_index = {}
```
You can get rid of wordiness by just making instruction_queues a dictionary

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+            if uid in uid_to_index:
+                instruction_queues[uid_to_index[uid]].put(("subscribe", args))
+
+class DeviceWriteThread(threading.Thread):
```
There is no need to extend the threading class.[http://stackoverflow.com/questions/660961/overriding-python-threading-thread-run]

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+    fake_device_queues = [queue.Queue() for _ in ports]
+
+    # each device has one write thread that receives instructions from the main thread and writes to devices
+    write_threads = [DeviceWriteThread(ser, iq, fake_device_queue) for ser, iq, fake_device_queue in zip(serials, instruction_queues, fake_device_queues)]
```
iq is a really hard to parse name. (so is ser)

Extra characters in variable names costs nothing but improves readability.

nikitakit commented:
No more big issues jump out to me on a second read-through. However, I think the possibility of plugging in last year's devices into the robot should not be ignored. Is there any way we can at least ensure that the protocol mismatch won't crash hibike? (Even better would be to report something like an "incompatible device" type to dawn, to make it very clear what the problem is)

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+
+            self.stateQueue.put(res)
+
+class FakeSubscriptionThread(threading.Thread):
```
No need to extend threading class.

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+            while True:
+                if self.quit:
+                    return
+                time.sleep(self.delay / 1000.0)
```
potentially inaccurate, as the number of processes/ threads running on the beaglebone scales up.
May not be an issue at all in practice, though.

Using signals wouldnt work here. Signals + threads leads to bad things in python.

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+                params_and_values = {}
+                for param, param_type in zip(self.params, param_types):
+                    if param_type in ("bool", ):
+                        params_and_values[param] = random.choice([True, False])
```
So the fake device thread return random values? 
How is this even tested for correctness?

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,142 @@
+        super().__init__()
+
+    def run(self):
+        if self.delay != 0:
```
Unsubscribing is not tested, right?

It looks like subscribing to a device starts a new thread that runs until told to quit. It doesn't look like it handles polling.

karthik-shanmugam replied:
This only supports subscribing. Unsubscribing in hibike is just subscribing with a delay of 0, which this also supports.

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,144 @@
+    newProcess = multiprocessing.Process(target=hibike_process, name="hibike_sim", args=[badThingsQueue, stateQueue, pipeFromChild])
+    newProcess.daemon = True
+    newProcess.start()
+    pipeToChild.send(["ready", []])
```
remove

yizow reviewed (changes requested):
This commit is massive. It should be broken up into many different commits, on many different branches, and you should make many different pull requests. There are very clear feature distinctions that should make the separate branches logically clear. 

Please make sure to write a full commit message. Possibly squash a lot of your bug fixing commits together before you pull request.

vdonato commented on hibike/README.md:
```diff
@@ -46,265 +104,271 @@ Each Smart Device will be assigned an 88-bit UID with the following data.
+|  0x09   | ColorSensor    |              |            |            |       |        |
+|  0x10   | DistanceSensor |              |            |            |       |        |
+|  0x11   | MetalDetector  |              |            |            |       |        |
+|  0xFFFF | ExampleDevice  | 0            | kumiko     | bool       | yes   | yes    |
```
kyaa~

karthik-shanmugam committed:
remove deprecated ready command from simulator
